packages/venia-concept/src/components/ProductImageCarousel/carousel.js

```diff
@@ -16,121 +16,103 @@ const ChevronIcons = {
+                >
+                    <Icon src={ChevronIcons[direction]} size={40} />
+                </button>
+            );
+        },
+        [handleLeftChevron, handleRightChevron]
     );
 
-    get sortedImages() {
-        const { images } = this.props;
-        return this.sortAndFilterImages(images);
-    }
-
-    leftChevronHandler = () => {
-        const sortedImages = this.sortedImages;
-        const { activeItemIndex } = this.state;
-        activeItemIndex > 0
-            ? this.updateActiveItemIndex(activeItemIndex - 1)
-            : this.updateActiveItemIndex(sortedImages.length - 1);
-    };
-
-    rightChevronHandler = () => {
-        const sortedImages = this.sortedImages;
-        const { activeItemIndex } = this.state;
-        this.updateActiveItemIndex((activeItemIndex + 1) % sortedImages.length);
-    };
-
-    getChevron = direction => (
-        <button
-            onClick={this[`${direction}ChevronHandler`]}
-            className={this.props.classes[`chevron-${direction}`]}
-        >
-            <Icon src={ChevronIcons[direction]} size={40} />
-        </button>
+    const placeholderImage = !currentImageLoaded && (
+        <img
+            className={classes.currentImage}
+            src={transparentPlaceholder}
+            alt={alt}
+        />
     );
 
-    setCurrentImageLoaded = () => {
-        this.setState({
-            currentImageLoaded: true
-        });
-    };
-
-    render() {
-        const { classes } = this.props;
-
-        const sortedImages = this.sortedImages;
-
-        const mainImage = sortedImages[this.state.activeItemIndex] || {};
-        const src = mainImage.file
-            ? resourceUrl(mainImage.file, { type: 'image-product', width: 640 })
-            : transparentPlaceholder;
-        const alt = mainImage.label || 'image-product';
-        return (
-            <div className={classes.root}>
-                <div className={classes.imageContainer}>
-                    {this.getChevron('left')}
-                    <img
-                        onLoad={this.setCurrentImageLoaded}
-                        className={classes.currentImage}
-                        src={src}
-                        alt={alt}
-                    />
-                    {!this.state.currentImageLoaded ? (
-                        <img
-                            className={classes.currentImage}
-                            src={transparentPlaceholder}
-                            alt={alt}
-                        />
-                    ) : null}
-                    {this.getChevron('right')}
-                </div>
-                <ThumbnailList
-                    items={sortedImages}
-                    activeItemIndex={this.state.activeItemIndex}
-                    updateActiveItemIndex={this.updateActiveItemIndex}
+    const handleImageLoad = useCallback(() => setCurrentImageLoaded(true));
+
+    // TODO: Evaluate why the re-render is occurring.
```

Comment:
A big thing I noticed, and maybe not because of this code, is that we are requesting every thumbnail/image on each click.

Reply:
Looks like the `setCurrentImageLoaded` code from before this refactor only ever ran for the first image load and was never set back to `false`.

https://github.com/magento-research/pwa-studio/blob/develop/packages/venia-concept/src/components/ProductImageCarousel/carousel.js#L92